utils/src/main/java/software/amazon/awssdk/utils/cache/CachedSupplier.java

```diff
@@ -156,54 +188,117 @@ private void refreshCache() {
                 // Make sure the value was not refreshed while we waited for the lock.
                 if (cacheIsStale() || shouldInitiateCachePrefetch()) {
 
+                    // It wasn't, call the supplier to update it.
+
                     if (prefetchStrategyInitialized.compareAndSet(false, true)) {
                         prefetchStrategy.initializeCachedSupplier(this);
                     }
 
-                    // It wasn't, call the supplier to update it.
-                    cachedValue = prefetchStrategy.fetch(valueSupplier);
+                    try {
+                        cachedValue = handleFetchedSuccess(prefetchStrategy.fetch(valueSupplier));
+                    } catch (RuntimeException t) {
+                        cachedValue = handleFetchFailure(t);
+                    }
                 }
             } finally {
                 if (lockAcquired) {
                     refreshLock.unlock();
                 }
             }
         } catch (InterruptedException e) {
-            handleInterruptedException("Interrupted waiting to refresh the value.", e);
+            Thread.currentThread().interrupt();
+            throw new IllegalStateException("Interrupted waiting to refresh a cached value.", e);
         }
     }
 
-    private void handleInterruptedException(String message, InterruptedException cause) {
-        Thread.currentThread().interrupt();
-        throw new IllegalStateException(message, cause);
+    /**
+     * Perform necessary transformations of the successfully-fetched value based on the stale value behavior of this supplier.
+     */
+    private RefreshResult<T> handleFetchedSuccess(RefreshResult<T> fetch) {
+        consecutiveStaleRetrievalFailures.set(0);
+
+        Instant now = clock.instant();
+
+        if (now.isBefore(fetch.staleTime())) {
+            return fetch;
+        }
+
+        switch (staleValueBehavior) {
+            case STRICT:
+                Instant newStale = now.plusSeconds(1);
+                log.warn(() -> "Retrieved value expiration is in the past (" + fetch.staleTime() + "). Using expiration "
+                               + "of " + newStale);
+                return fetch.toBuilder().staleTime(newStale).build(); // Refresh again in 1 second
+            case ALLOW:
+                Instant newStaleTime = jitterTime(now, Duration.ofMinutes(1), Duration.ofMinutes(10));
+                log.warn(() -> "Cached value expiration has been extended to " + newStaleTime + " because the downstream "
+                               + "service returned a time in the past: " + fetch.staleTime());
+
+                return fetch.toBuilder()
+                            .staleTime(newStaleTime)
+                            .build();
```

Comment:
Static stability! IMDS returned an expired value. Use that value for the next 1-10 minutes before trying to get a new value from IMDS.